seamm_util/plotting.py

```diff
@@ -0,0 +1,554 @@
+
+        Parameters
+        ----------
+        key : str
+            The key to fetch.
+        default : any value, optional
+            The value to return if the key is not in the data. Defaults to an
+            empty string.
+
+        Returns
+        -------
+        any
+            Returns the value associated with the key, or the default.
+        """
+
+        return self.data.get(key, default)
+
+    def add_plot(self, name, **kwargs):
+        """Create a new plot and return it.
+
+        Parameters
+        ----------
+        name : str
+            The name of the plot, which must be unique.
+
+        Returns
+        -------
+        The plot object.
+        """
+        if name in self._plots:
+            raise KeyError("Plot '{}' already exists.".format(name))
+
+        self._plots[name] = Plot(**kwargs)
+
+        return self._plots[name]
+
+    def get_plot(self, name):
+        """Return the named plot.
+
+        Parameters
+        ----------
+        name : str
+            The name of the plot, which must exist.
+
+        Returns
+        -------
+        The plot object.
+        """
+        if name not in self._plots:
+            raise KeyError("Plot '{}' does not exist.".format(name))
+
+        return self._plots[name]
+
+    def grid_plots(self, *args, **kwargs):
+        """Define the layout of the plots.
+
+        When there is more than one plot, they need to be laid out in a grid.
+        This method is called to define the layout, and will follow closely the
+        approach used in Tk grid.
+
+        Parameters
+        ----------
+        *args
+            Positional arguments specifying the layout of one or more rows.
+            each row is given as a string composed of the following:
+
+                <plot> :
+                     is the name of a plot.
+
+                x :
+                    leaves an empty column between the plot on the
+                    left and the plot on the right.
+
+                ^ :
+                    extends the row span of the plot above the ^'s in the
+                    grid. The number of ^'s in a row must match the number
+                    of columns spanned by the plot above it.
+
+                - :
+                    increases the column span of the plot to the
+                    left. Several -'s in a row will successively increase
+                    the number of columns spanned. A - may not follow a ^
+                    or a x, nor may it be the first plot argument to
+                    grid configure.
+
+        **kwargs
+            Keyword arguments for options for the given rows:
+                padx : float, optional
+                    The fraction of space to leave between plots in a row.
+                    Defaults to 0.02.
+                pady : float, optional
+                    The fraction of space to leave between rows. Defaults to
+                    0.02.
```

Comment:
I'm having trouble see how the `args` and `kwargs` work. An example section in the docstring may be helpful here, or perhaps we will just want to have a detailed doc page/demo online. I'm not seeing a unit test besides testing errors, so I'm not sure how to create a working example. If it's based on an approach followed by TKGrid, might we also link to that documentation?

Reply:
I think we need a more general documentation, a user guide for plotting. Looking at it function by function is difficult and confusing.

There are two sets of tests, test_plotting_figure and test_plotting. The first does simple testing and only tests errors in gridding. The second creates complete graphs as HTML and compares against known HTML, and uses grid more elaborately. One nice think about having HTML test output is you can double click and see that the graph is what was expected.